### PR TITLE
Frequent apps in the launcher

### LINKED ISSUE #16: Frequent apps page in the launcher
This would be similar to the Frequent tab in the GNOME launcher and would show apps that user has used frequently/recently. Useful for users with a lot of apps installed.

### PR DESCRIPTION
**Pull request by iBeliever**
The algorithm is taken from GNOME Shell.

Fixes #16

### CODE REVIEW

File: declarative/launcher/usagetracker.cpp [hunk 1]
```diff
@@ -0,0 +1,203 @@
+/****************************************************************************
+ * This file is part of Hawaii Shell.
+ *
+ * Copyright (C) 2016 <sonrisesoftware@gmail.com>
```
Comment (plfiorini) [inline declarative/launcher/usagetracker.cpp]: name missing

Comment (iBeliever) [inline declarative/launcher/usagetracker.cpp]: Fixed.

File: declarative/launcher/appusage.cpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ *
+ * $END_LICENSE$
+ ***************************************************************************/
+
+#include "appusage.h"
+
+#include <QtCore/QVariant>
+#include <QtSql/QSqlQuery>
+
+AppUsage::AppUsage(const QString &appId, QObject *parent) : QObject(parent), appId(appId)
```
Comment (plfiorini) [inline declarative/launcher/appusage.cpp]: nitpick: align like Qt:

AppUsage::AppUsage(const QString &appId, QObject *parent)
    : QObject(parent)
    , appId(appId)

**Comment by plfiorini** [comment]
at some point it prints this and dies:
QSqlDatabasePrivate::database: unable to open database: "out of memory Error opening database"

**Comment by iBeliever** [comment]
@plfiorini Oh, right, I forgot to fix that. The database won't create the data directory, so it throws that very unhelpful error. You need to create the directory `~/.local/share/Hawaii/Hawaii` if you want to test it. I'll fix that this evening after work.

**Comment by iBeliever** [comment]
@plfiorini Here are three different potential designs:

* **First page in launcher pages, special icon for indicator:**
  hawaii_012

* **Frequent/all apps toggle in place of the categories toggle:**

   hawaii_014


* **Same as above, only two rows (all apps grid would still have four rows):**
   
   hawaii_013

 
Any preferences or suggestions for something different?

**Comment by plfiorini** [comment]
I wouldn't remove the category list and just add a button to show the most recent apps in 2 rows and no pages (that makes it for 12 frequent apps at most).

File: declarative/launcher/appusage.cpp (hunk 2 above)
Comment (iBeliever) [inline declarative/launcher/appusage.cpp]: Fixed.

**Comment by iBeliever** [comment]
@plfiorini Ok, I've updated the launcher popup to go with that design.

**Comment by iBeliever** [comment]
Go ahead and review now.

**Comment by plfiorini** [comment]
@iBeliever  it doesn't show anything

there's an error: file:///home/plfiorini/git/qmlos/shell/shell/launcher/LauncherPopOver.qml:117: ReferenceError: categories is not defined

**Comment by plfiorini** [comment]
@iBeliever Now it starts up.

Here's my db:

org.hawaiios.PolkitAgent|1471793475|3
gnome-calendar|1471793967|31
dconf-editor|1471794017|5

- Policy Kit agent is saved in the db, you should not save apps with NoDisplay=true
- Calendar didn't appear in the frequent apps page although i started it many times and left focused for more than 7 seconds
- dconf-editor doesn't appear either
- When I was using the desktop the only frequent app that appeared (color profile) is not in the db.
- Restarting the desktop there are no frequent apps

**Comment by plfiorini** [comment]
This is the db now:

gnome-calendar|1471793967|31
dconf-editor|1471794017|5
org.hawaiios.PolkitAgent|1471794322|4
gcm-viewer|1471794372|7
file-roller|1471794396|3

Finally gcm-viewer got saved. It has a score of 7 and it shows up but not Calendar which is higher

**Comment by iBeliever** [comment]
@plfiorini All the issues you found should be addressed now. The issue with apps not showing in the view was that the app IDs set by the launcher didn't match any desktop files. I've fixed that with a temporary workaround [here](https://github.com/qmlos/libqmlos/blob/develop/src/core/desktopfile.cpp#L36).

**Comment by plfiorini** [comment]
ASSERT: "found != nullptr" in file /home/plfiorini/git/qmlos/shell/declarative/launcher/launchermodel.cpp, line 338

**Comment by iBeliever** [comment]
@plfiorini Fixed the assert.

**Comment by plfiorini** [comment]
@iBeliever works, you can merge!

**Comment by iBeliever** [comment]
Squashed and merged.

**Comment by plfiorini** [comment]
:+1:

**Pull request closed without merging**